### DISCUSSION TITLE
StdoutCallback() deprecation - alternative method

### DISCUSSION THREAD

**Discussion by tos-valantic** [sinaptik-ai/pandas-ai#775] [Q&A] [answered]
Hi, 
in pandasai==1.4.7 I have used the now deprecated StdoutCallback() to catch the llm-generated function and display it in a flask app. Is there any alternative to get the function? explain() or rephrase_query() did not help with that.
Any help is welcome!

```# Custom stdout capturing:
@contextmanager
def capture_stdout():
    old_stdout = sys.stdout
    sys.stdout = mystdout = StringIO()
    yield mystdout
    sys.stdout = old_stdout
    
    
pandasai_sdf = SmartDataframe(
    df_pandas,
    config={"llm": llm, "verbose": False, "callback": StdoutCallback()},
)

prompt_extension = "\nPrevent devide by zero."
with capture_stdout() as stdout_buffer:
    result = pandasai_sdf.chat(
        query=user_input + prompt_extension, output_type="dataframe"
    )

stdout_content = stdout_buffer.getvalue()
```

**Comment by gventuri** [comment, marked as answer]
@tos-wdp, If the goal is to get the last code executed, you can simply use `df.last_code_executed` or `lake.last_code_executed` where `df` and `lake` are either the variable of your SmartDataframe or SmartDatalake.

**Reply by tos-valantic** [reply]
Thank you, this is a nice solution!

Minor downside is that the comments are missing.

stdout_content:
![stdout_content](https://github.com/gventuri/pandas-ai/assets/126463646/2ce822ed-ea6b-46b5-98d4-ba1c97a467f5)

last_code_executed:
![last_code_executed](https://github.com/gventuri/pandas-ai/assets/126463646/72c0037b-5bd7-48e4-97f0-c9bc991153df)
